File: weather-app/app/src/main/java/com/rperazzo/weatherapp/viewmodel/WeatherViewModel.java
package com.rperazzo.weatherapp.viewmodel;

import android.app.Application;
import android.arch.lifecycle.AndroidViewModel;
import android.arch.lifecycle.LiveData;
import android.arch.lifecycle.MutableLiveData;
import android.content.Context;
import android.content.SharedPreferences;
import android.os.AsyncTask;
import android.support.annotation.NonNull;

import com.rperazzo.weatherapp.services.WeatherService;
import com.rperazzo.weatherapp.utils.WeatherManager;
import com.rperazzo.weatherapp.model.FindResult;
import com.rperazzo.weatherapp.utils.Const;

import java.io.IOException;

public class WeatherViewModel extends AndroidViewModel {

    private SharedPreferences sp;

    private WeatherService weatherService;

    private MutableLiveData<FindResult> weather = new MutableLiveData<FindResult>();

    public WeatherViewModel(@NonNull Application application) {
        super(application);

        this.weatherService = WeatherManager.getService();
        this.sp = application.getSharedPreferences(Const.PREFERENCE_NAME, Context.MODE_PRIVATE);
    }

    public LiveData<FindResult> find() {
        return weather;
    }

    public LiveData<FindResult> find(final String query) {
        final String unit = getTemperatureUnit();

        AsyncTask async = new AsyncTask<Void, Void, Void>() {

            @Override
            protected Void doInBackground(Void... voids) {

                try {

                    FindResult result = weatherService
                            .find(query, unit, Const.APP_KEY)
                            .execute()
                            .body();

                    weather.postValue(result);

                } catch (IOException e) {
                    e.printStackTrace();
                }

                return null;
            }

        }.execute();

        return weather;
    }

    public void saveTemperatureUnit(String newUnit) {
        sp
                .edit()
                .putString(Const.TEMPERATURE_UNIT_KEY, newUnit)
                .apply();
    }

    public String getTemperatureUnit() {
        return sp.getString(Const.TEMPERATURE_UNIT_KEY, "metric");
    }

}

File: weather-app/app/src/main/java/com/rperazzo/weatherapp/MVVM/viewmodel/WeatherViewModel.kt
package com.rperazzo.weatherapp.MVVM.viewmodel

import android.app.Application
import android.arch.lifecycle.AndroidViewModel
import android.arch.lifecycle.LiveData
import android.arch.lifecycle.MutableLiveData
import android.content.Context
import android.content.SharedPreferences
import com.rperazzo.weatherapp.MVVM.model.entity.FindResult
import com.rperazzo.weatherapp.MVVM.model.service.WeatherManager
import com.rperazzo.weatherapp.MVVM.model.service.WeatherService
import com.rperazzo.weatherapp.MVVM.utils.Constantes.*


class WeatherViewModel(application: Application) : AndroidViewModel(application) {
    private var sp: SharedPreferences? = null

    private var weatherService: WeatherService? = null

    private val weather = MutableLiveData<FindResult>()

    fun WeatherViewModel(application: Application) {
        weatherService = WeatherManager.getService()
        sp = application.getSharedPreferences(PREFERENCE_NAME, Context.MODE_PRIVATE)
    }

    fun find(): LiveData<FindResult>? {
        return weather
    }

    fun find(query: String?): LiveData<FindResult>? {
        /* val unit = getTemperatureUnit()
        object : AsyncTask<Void?, Void?, Void?>() {
            protected fun doInBackground(vararg voids: Void): Void? {
                try {
                    val result: FindResult? = query?.let {
                        weatherService
                                .find(it, unit, API_KEY)
                                .execute()
                                .body()
                    }
                    weather.postValue(result)
                } catch (e: IOException) {
                    e.printStackTrace()
                }
                return null
            }
        }.execute() */
        return weather
    }

    fun saveTemperatureUnit(newUnit: String?) = sp
        ?.edit()
        ?.putString(TEMPERATURE_UNIT_KEY, newUnit)
        ?.apply()

    fun getTemperatureUnit(): String {
        return sp!!.getString(TEMPERATURE_UNIT_KEY, "metric")
    }

}